File: packages/ik_rules/test/trackers_test.dart
import 'package:ik_content/ik_content.dart';
import 'package:ik_parity/ik_parity.dart';
import 'package:ik_rules/ik_rules.dart';
import 'package:test/test.dart';

void main() {
  late GameDatabase db;
  late PlayerSave fresh;

  setUpAll(() {
    db = filterLaunchContent(assertGameDatabaseShape(contentDatabaseJson()));
  });

  setUp(() {
    fresh = createNewSave(db, 1_000);
  });

  ActionRow action(String id) => db.actions.firstWhere((row) => row.actionId == id);

  EnemyRow enemy(String id) => db.enemies.firstWhere((row) => row.enemyId == id);

  test('cow and bull open separate loot sections', () {
    final tracking = resumeAllTrackers(fresh, 1_000);
    final cow = applyCombatVictory(
      db,
      tracking,
      action('ACN-0001'),
      enemy('ENM-0001'),
      () => 0,
      2_000,
    );
    final both = applyCombatVictory(
      db,
      cow.save,
      action('ACN-0002'),
      enemy('ENM-0002'),
      () => 0,
      3_000,
    );
    expect(both.save.lootTrackers.keys, containsAll(['enemy:ENM-0001', 'enemy:ENM-0002']));
    expect(both.save.lootTrackers['enemy:ENM-0001']!.completions, 1);
    expect(both.save.lootTrackers['enemy:ENM-0002']!.completions, 1);
  });

  test('resetting cows leaves the bull section', () {
    final tracking = resumeAllTrackers(fresh, 1_000);
    final cow = applyCombatVictory(
      db,
      tracking,
      action('ACN-0001'),
      enemy('ENM-0001'),
      () => 0,
      2_000,
    );
    final both = applyCombatVictory(
      db,
      cow.save,
      action('ACN-0002'),
      enemy('ENM-0002'),
      () => 0,
      3_000,
    );
    final reset = resetLootTracker(both.save, 'enemy:ENM-0001');
    expect(reset.lootTrackers.containsKey('enemy:ENM-0001'), isFalse);
    expect(reset.lootTrackers.containsKey('enemy:ENM-0002'), isTrue);
  });

  test('gathering starts an action loot section and XP tracker', () {
    final mine = action('ACN-0018');
    final completed = completeGatheringAction(
      db,
      resumeAllTrackers(fresh, 1_000),
      mine,
      () => 0,
      4_000,
    );
    expect(completed.save.lootTrackers.containsKey('action:ACN-0018'), isTrue);
    expect(completed.save.xpTrackers.containsKey('SKL-0002'), isTrue);
    expect(completed.save.xpTrackers.containsKey(totalXpTrackerId), isTrue);
  });

  test('standard production awards XP but not loot items', () {
    var save = addItemsToInventory(fresh, 'ITEM-0025', 10).save;
    save = resumeAllTrackers(save, 1_000).copyWith(currentLocationId: 'LOC-0023');
    final queued = beginProductionQueue(db, save, 'ACT-0017', 'RCP-0001', 1, 1_000);
    expect(queued.ok, isTrue);
    final finished = completeProductionCraft(db, queued.save!, 5_000, () => 0);
    expect(finished, isNotNull);
    expect(finished!.save.lootTrackers, isEmpty);
    expect(finished.save.xpTrackers.containsKey(totalXpTrackerId), isTrue);
  });

  test('new saves start with loot and XP tracking off', () {
    expect(lootTrackersPaused(fresh), isTrue);
    expect(xpTrackersPaused(fresh), isTrue);
    final mined = completeGatheringAction(db, fresh, action('ACN-0018'), () => 0, 4_000);
    expect(mined.save.lootTrackers.containsKey('action:ACN-0018'), isFalse);
    expect(mined.save.xpTrackers.containsKey('SKL-0002'), isFalse);
  });

  test('loot and XP pause independently', () {
    var save = pauseLootTrackers(resumeAllTrackers(fresh, 1_000), 1_000);
    final mined = completeGatheringAction(db, save, action('ACN-0018'), () => 0, 4_000);
    expect(mined.save.lootTrackers.containsKey('action:ACN-0018'), isFalse);
    expect(mined.save.xpTrackers.containsKey('SKL-0002'), isTrue);

    save = resumeLootTrackers(pauseXpTrackers(mined.save, 4_000), 4_000);
    final again = completeGatheringAction(db, save, action('ACN-0018'), () => 0, 6_000);
    expect(again.save.lootTrackers.containsKey('action:ACN-0018'), isTrue);
    expect(
      again.save.xpTrackers['SKL-0002']!.xpGained,
      mined.save.xpTrackers['SKL-0002']!.xpGained,
    );
  });

  test('xp/hr uses elapsed time since the tracker started', () {
    final entry = XpTrackerEntry(skillId: 'SKL-0001', startedAtMs: 0, xpGained: 3_600);
    expect(xpPerHour(entry, 3_600_000), 3_600);
  });

  test('off freezes xp/hr and on excludes the paused window', () {
    final runningSave = resumeAllTrackers(fresh, 1_000);
    final mined = completeGatheringAction(db, runningSave, action('ACN-0018'), () => 0, 4_000);
    final running = xpPerHour(mined.save.xpTrackers['SKL-0002']!, 10_000, mined.save);
    final paused = pauseXpTrackers(mined.save, 10_000);
    expect(xpTrackersPaused(paused), isTrue);
    expect(xpPerHour(paused.xpTrackers['SKL-0002']!, 20_000, paused), running);
    final resumed = resumeXpTrackers(paused, 20_000);
    expect(xpTrackersPaused(resumed), isFalse);
    expect(xpPerHour(resumed.xpTrackers['SKL-0002']!, 20_000, resumed), running);
  });
}

File: src/game/trackers/trackers.test.ts
import { readFileSync } from 'node:fs'
import { resolve } from 'node:path'
import { describe, expect, it } from 'vitest'
import { applyCombatVictory } from '../combat/engine'
import { completeGatheringAction } from '../activity/engine'
import { beginProductionQueue, completeProductionCraft } from '../production/engine'
import { prepareDatabase } from '../data/loadDatabase'
import { createNewSave } from '../save/saveStore'
import { addItemsToInventory } from '../activity/rewards'
import { TOTAL_XP_TRACKER_ID } from '../save/types'
import { pauseTrackers, resetLootTracker, resumeTrackers, trackersPaused, xpPerHour } from './trackers'

const rawDatabase = JSON.parse(
  readFileSync(resolve(process.cwd(), 'content/data/game-database.json'), 'utf8'),
)

describe('trackers', () => {
  const { launch } = prepareDatabase(rawDatabase)
  const action = (id: string) => launch.Actions.find((row) => row['Action ID'] === id)!
  const enemy = (id: string) => launch.Enemies.find((row) => row['Enemy ID'] === id)!

  it('opens separate loot sections for cow and bull', () => {
    const fresh = createNewSave(launch, 1000)
    const cow = applyCombatVictory(launch, fresh, action('ACN-0001'), enemy('ENM-0001'), () => 0, 2000)
    const both = applyCombatVictory(launch, cow.save, action('ACN-0002'), enemy('ENM-0002'), () => 0, 3000)
    expect(Object.keys(both.save.lootTrackers)).toEqual(
      expect.arrayContaining(['enemy:ENM-0001', 'enemy:ENM-0002']),
    )
  })

  it('resetting cows leaves the bull section', () => {
    const fresh = createNewSave(launch, 1000)
    const cow = applyCombatVictory(launch, fresh, action('ACN-0001'), enemy('ENM-0001'), () => 0, 2000)
    const both = applyCombatVictory(launch, cow.save, action('ACN-0002'), enemy('ENM-0002'), () => 0, 3000)
    const reset = resetLootTracker(both.save, 'enemy:ENM-0001')
    expect(reset.lootTrackers['enemy:ENM-0001']).toBeUndefined()
    expect(reset.lootTrackers['enemy:ENM-0002']).toBeDefined()
  })

  it('starts an action loot section when mining copper', () => {
    const fresh = createNewSave(launch, 1000)
    const completed = completeGatheringAction(launch, fresh, action('ACN-0018'), () => 0, 4000)
    expect(completed.save.lootTrackers['action:ACN-0018']).toBeDefined()
    expect(completed.save.xpTrackers['SKL-0002']).toBeDefined()
    expect(completed.save.xpTrackers[TOTAL_XP_TRACKER_ID]).toBeDefined()
  })

  it('does not track standard production outputs as loot', () => {
    let save = addItemsToInventory(createNewSave(launch, 1000), 'ITEM-0025', 10).save
    save = { ...save, currentLocationId: 'LOC-0023' }
    const queued = beginProductionQueue(launch, save, 'ACT-0017', 'RCP-0001', 1, 1000)
    expect(queued.ok).toBe(true)
    if (!queued.ok) return
    const finished = completeProductionCraft(launch, queued.save, 5000, () => 0)
    expect(finished).not.toBeNull()
    expect(Object.keys(finished!.save.lootTrackers)).toEqual([])
    expect(finished!.save.xpTrackers[TOTAL_XP_TRACKER_ID]).toBeDefined()
  })

  it('starts stopped, with xp/hr at 0 until Start', () => {
    const fresh = createNewSave(launch, 1000)
    expect(trackersPaused(fresh)).toBe(true)
    const mined = completeGatheringAction(launch, fresh, action('ACN-0018'), () => 0, 4000)
    expect(trackersPaused(mined.save)).toBe(true)
    expect(xpPerHour(mined.save.xpTrackers['SKL-0002']!, 10_000, mined.save)).toBe(0)
    const started = resumeTrackers(mined.save, 10_000)
    expect(trackersPaused(started)).toBe(false)
    expect(started.xpTrackers['SKL-0002']!.startedAtMs).toBe(10_000)
    const gained = started.xpTrackers['SKL-0002']!.xpGained
    expect(xpPerHour(started.xpTrackers['SKL-0002']!, 10_000 + 3_600_000, started)).toBe(gained)
  })

  it('computes xp/hr from elapsed time', () => {
    expect(xpPerHour({ skillId: 'SKL-0001', startedAtMs: 0, xpGained: 3600 }, 3_600_000)).toBe(3600)
  })

  it('freezes xp/hr while stopped and excludes paused time after start', () => {
    const runningSave = resumeTrackers(createNewSave(launch, 1000), 1000)
    const mined = completeGatheringAction(launch, runningSave, action('ACN-0018'), () => 0, 4000)
    const running = xpPerHour(mined.save.xpTrackers['SKL-0002']!, 10_000, mined.save)
    const paused = pauseTrackers(mined.save, 10_000)
    expect(trackersPaused(paused)).toBe(true)
    expect(xpPerHour(paused.xpTrackers['SKL-0002']!, 20_000, paused)).toBe(running)
    const resumed = resumeTrackers(paused, 20_000)
    expect(trackersPaused(resumed)).toBe(false)
    expect(xpPerHour(resumed.xpTrackers['SKL-0002']!, 20_000, resumed)).toBe(running)
  })
})
